caddyhttp/proxy/proxy.go

```diff
@@ -278,7 +278,10 @@ func mutateHeadersByRules(headers, rules http.Header, repl httpserver.Replacer)
 		} else if strings.HasPrefix(ruleField, "-") {
 			headers.Del(strings.TrimPrefix(ruleField, "-"))
 		} else if len(ruleValues) > 0 {
-			headers.Set(ruleField, repl.Replace(ruleValues[len(ruleValues)-1]))
+			var replacement = repl.Replace(ruleValues[len(ruleValues)-1])
+			if len(replacement) > 0 {
+				headers.Set(ruleField, replacement)
+			}
```

Comment:
I have an alternative idea to implement this. How about to define an extra prefix, let's say `?` and apply conditional filtering only on headers marked with this. This way it wouldn't change existing behavior, and people have issues with empty headers can use this method e.g.:

`header_upstream ?Upgrade {>Upgrade}`